In [96]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf
import os
import markdown

Step 1: Processing the Data

1.1 Load the Dataset
Here I start by importing the necessary libraries and loading the dataset from the provided cloud URL.

In [79]:
# Load the dataset
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")

# Display the first few rows
application_df.head()


,EIN,NAME,APPLICATION_TYPE,AFFILIATION,CLASSIFICATION,USE_CASE,ORGANIZATION,STATUS,INCOME_AMT,SPECIAL_CONSIDERATIONS,ASK_AMT,IS_SUCCESSFUL
0,10520599,BLUE KNIGHTS MOTORCYCLE CLUB,T10,Independent,C1000,ProductDev,Association,1,0,N,5000,1
1,10531628,AMERICAN CHESAPEAKE CLUB CHARITABLE TR,T3,Independent,C2000,Preservation,Co-operative,1,1-9999,N,108590,1
2,10547893,ST CLOUD PROFESSIONAL FIREFIGHTERS,T5,CompanySponsored,C3000,ProductDev,Association,1,0,N,5000,0
3,10553066,SOUTHSIDE ATHLETIC ASSOCIATION,T3,CompanySponsored,C2000,Preservation,Trust,1,10000-24999,N,6692,1
4,10556103,GENETIC RESEARCH INSTITUTE OF THE DESERT,T3,Independent,C1000,Heathcare,Trust,1,100000-499999,N,142590,1


In [80]:
print(application_df.columns)

Index(['EIN', 'NAME', 'APPLICATION_TYPE', 'AFFILIATION', 'CLASSIFICATION',
       'USE_CASE', 'ORGANIZATION', 'STATUS', 'INCOME_AMT',
       'SPECIAL_CONSIDERATIONS', 'ASK_AMT', 'IS_SUCCESSFUL'],
      dtype='object')


1.2 Identify Target and Feature Variables
Here I determine which column are the target variable and which are features.

Target Variable

IS_SUCCESSFUL: This is my binary target (0 or 1), indicating whether the organization successfully used the funding.

Feature Variables

All other columns (except EIN and NAME) are features that provide metadata about the organization.


In [81]:
# Identify target and feature variables
target_variable = "IS_SUCCESSFUL"

feature_variables = application_df.columns.drop(["EIN", "NAME", target_variable]).tolist()

print("Target Variable:", target_variable)

print("Feature Variables:", feature_variables)


Target Variable: IS_SUCCESSFUL
Feature Variables: ['APPLICATION_TYPE', 'AFFILIATION', 'CLASSIFICATION', 'USE_CASE', 'ORGANIZATION', 'STATUS', 'INCOME_AMT', 'SPECIAL_CONSIDERATIONS', 'ASK_AMT']


1.3 Drop Irrelevant Columns
Next, the columns EIN and NAME are identification variables and do not contribute to prediction.

Here I remove these columns using .drop(["EIN", "NAME"], axis=1). The dataset is now cleaner and focused only on relevant features.

In [82]:
# Drop the EIN and NAME columns
application_df = application_df.drop(["EIN", "NAME"], axis=1)

# Verify drop
application_df.head()

,APPLICATION_TYPE,AFFILIATION,CLASSIFICATION,USE_CASE,ORGANIZATION,STATUS,INCOME_AMT,SPECIAL_CONSIDERATIONS,ASK_AMT,IS_SUCCESSFUL
0,T10,Independent,C1000,ProductDev,Association,1,0,N,5000,1
1,T3,Independent,C2000,Preservation,Co-operative,1,1-9999,N,108590,1
2,T5,CompanySponsored,C3000,ProductDev,Association,1,0,N,5000,0
3,T3,CompanySponsored,C2000,Preservation,Trust,1,10000-24999,N,6692,1
4,T3,Independent,C1000,Heathcare,Trust,1,100000-499999,N,142590,1


1.4 Count Unique Values in Each Column
Here I will check how many unique values exist in each categorical column.

Why is this Important?

Columns with too many unique values may need encoding.
Rare categories should be combined into an "Other" category.

In [83]:
# Count unique values
unique_counts = application_df.nunique()
print(unique_counts)

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64


1.5 Combine Rare Categories in Categorical Columns
Next, I will analyze the APPLICATION_TYPE and CLASSIFICATION columns to identify rare categories.

Handling APPLICATION_TYPE

In [84]:
# Count occurrences of each application type
application_counts = application_df['APPLICATION_TYPE'].value_counts()

# Identify rare application types (less than 500 occurrences)
application_types_to_replace = list(application_counts[application_counts < 500].index)

# Replace rare categories with "Other"
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app, "Other")

# Verify changes
print(application_df['APPLICATION_TYPE'].value_counts())

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64


Handling CLASSIFICATION

In [85]:
# Count occurrences of each classification type
class_counts = application_df['CLASSIFICATION'].value_counts()

# Identify rare classifications (less than 1000 occurrences)
classifications_to_replace = list(class_counts[class_counts < 1000].index)

# Replace rare categories with "Other"
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls, "Other")

# Verify changes
print(application_df['CLASSIFICATION'].value_counts())

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: count, dtype: int64


Explanation of why I did this: 
I count occurrences and replace categories with few instances.This helps prevent overfitting on rare categories.


1.6 Encode Categorical Variables
Neural networks require numerical inputs, so I convert categorical variables into numerical format using one-hot encoding

Explanation: 

-pd.get_dummies() converts categorical variables into separate binary columns.

-This ensures that all inputs are numeric.

In [86]:
# Perform one-hot encoding
application_with_dummies_df = pd.get_dummies(application_df)

# View column names
print(application_with_dummies_df.columns)

Index(['STATUS', 'ASK_AMT', 'IS_SUCCESSFUL', 'APPLICATION_TYPE_Other',
       'APPLICATION_TYPE_T10', 'APPLICATION_TYPE_T19', 'APPLICATION_TYPE_T3',
       'APPLICATION_TYPE_T4', 'APPLICATION_TYPE_T5', 'APPLICATION_TYPE_T6',
       'APPLICATION_TYPE_T7', 'APPLICATION_TYPE_T8',
       'AFFILIATION_CompanySponsored', 'AFFILIATION_Family/Parent',
       'AFFILIATION_Independent', 'AFFILIATION_National', 'AFFILIATION_Other',
       'AFFILIATION_Regional', 'CLASSIFICATION_C1000', 'CLASSIFICATION_C1200',
       'CLASSIFICATION_C2000', 'CLASSIFICATION_C2100', 'CLASSIFICATION_C3000',
       'CLASSIFICATION_Other', 'USE_CASE_CommunityServ', 'USE_CASE_Heathcare',
       'USE_CASE_Other', 'USE_CASE_Preservation', 'USE_CASE_ProductDev',
       'ORGANIZATION_Association', 'ORGANIZATION_Co-operative',
       'ORGANIZATION_Corporation', 'ORGANIZATION_Trust', 'INCOME_AMT_0',
       'INCOME_AMT_1-9999', 'INCOME_AMT_10000-24999',
       'INCOME_AMT_100000-499999', 'INCOME_AMT_10M-50M', 'INCOME_AMT_1M-5M

1.7 Split the Data into Training and Testing Sets
Here I define X (features) and y (target) and split the dataset. The train_test_split() function divides data into 75% training and 25% testing sets. This ensures the model learns on one set and is tested on unseen data.

In [87]:
# Define X (features) and y (target)
X = application_with_dummies_df.drop(["IS_SUCCESSFUL"], axis=1).values
y = application_with_dummies_df["IS_SUCCESSFUL"].values

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [88]:
X_train, X_test, y_train, y_test

(array([[1, 5000, False, ..., False, True, False],
        [1, 5000, False, ..., False, True, False],
        [1, 5000, False, ..., False, True, False],
        ...,
        [1, 103405, False, ..., False, True, False],
        [1, 5000, False, ..., False, True, False],
        [1, 1869240, False, ..., False, True, False]], dtype=object),
 array([[1, 14651, False, ..., False, True, False],
        [1, 5144674, False, ..., True, True, False],
        [1, 5000, False, ..., False, True, False],
        ...,
        [1, 5000, False, ..., False, True, False],
        [1, 5000, False, ..., False, True, False],
        [1, 5000, False, ..., False, True, False]], dtype=object),
 array([0, 0, 0, ..., 1, 0, 1]),
 array([0, 0, 1, ..., 0, 0, 1]))

1.8 Scale the Data
Neural networks perform better when data is scaled. Standardizing the data ensures all features have the same scale, improving model performance.


In [89]:
from sklearn.preprocessing import StandardScaler

# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the scaler on training data
X_scaler = scaler.fit(X_train)

# Transform the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

Step 2: Compile, Train, and Evaluate the Model

2.1 Define the Neural Network

Explanation

-Two hidden layers are used to capture complex patterns.

-ReLU activation is used in hidden layers for efficiency.

-Sigmoid activation in output ensures a probability output (binary classification).


In [90]:
# Define model parameters
number_input_features = len(X_train[0])
hidden_nodes_layer1 = 80
hidden_nodes_layer2 = 30

# Create the model
nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Summary
nn.summary()


Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_10 (Dense)            (None, 80)                3520      
                                                                 
 dense_11 (Dense)            (None, 30)                2430      
                                                                 
 dense_12 (Dense)            (None, 1)                 31        
                                                                 
Total params: 5,981
Trainable params: 5,981
Non-trainable params: 0
_________________________________________________________________


2.2 Compile and Train the Model

Explanation
Here I use binary cross-entropy for loss because it's a binary classification. The Adam optimizer helps efficiently update weights. The model runs for 100 epochs to improve performance.

In [91]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 1s 899us/step - loss: 0.5721 - accuracy: 0.7208
Epoch 2/100
804/804 [==============================] - 1s 871us/step - loss: 0.5565 - accuracy: 0.7290
Epoch 3/100
804/804 [==============================] - 1s 885us/step - loss: 0.5524 - accuracy: 0.7315
Epoch 4/100
804/804 [==============================] - 1s 914us/step - loss: 0.5512 - accuracy: 0.7313
Epoch 5/100
804/804 [==============================] - 1s 933us/step - loss: 0.5503 - accuracy: 0.7321
Epoch 6/100
804/804 [==============================] - 1s 936us/step - loss: 0.5486 - accuracy: 0.7336
Epoch 7/100
804/804 [==============================] - 1s 939us/step - loss: 0.5478 - accuracy: 0.7345
Epoch 8/100
804/804 [==============================] - 1s 947us/step - loss: 0.5477 - accuracy: 0.7345
Epoch 9/100
804/804 [==============================] - 1s 885us/step - loss: 0.5469 - accuracy: 0.7345
Epoch 10/100
804/804 [==============================] - 1s 882us/step - l

804/804 [==============================] - 1s 862us/step - loss: 0.5372 - accuracy: 0.7409
Epoch 80/100
804/804 [==============================] - 1s 877us/step - loss: 0.5372 - accuracy: 0.7417
Epoch 81/100
804/804 [==============================] - 1s 867us/step - loss: 0.5366 - accuracy: 0.7403
Epoch 82/100
804/804 [==============================] - 1s 889us/step - loss: 0.5369 - accuracy: 0.7410
Epoch 83/100
804/804 [==============================] - 1s 865us/step - loss: 0.5371 - accuracy: 0.7405
Epoch 84/100
804/804 [==============================] - 1s 892us/step - loss: 0.5365 - accuracy: 0.7407
Epoch 85/100
804/804 [==============================] - 1s 890us/step - loss: 0.5366 - accuracy: 0.7404
Epoch 86/100
804/804 [==============================] - 1s 882us/step - loss: 0.5367 - accuracy: 0.7408
Epoch 87/100
804/804 [==============================] - 1s 906us/step - loss: 0.5367 - accuracy: 0.7403
Epoch 88/100
804/804 [==============================] - 1s 862us/step - loss:

2.3 Evaluate Model Performance: Here I assess model accuracy and loss on unseen test data.



In [92]:
# Evaluate model
model_loss, model_accuracy = nn.evaluate(X_test_scaled, y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5569 - accuracy: 0.7247 - 228ms/epoch - 852us/step
Loss: 0.5568712949752808, Accuracy: 0.7246647477149963


2.4 Save the Model

In [93]:
# Save the model
nn.save('AlphabetSoupCharity.h5')

Step 3: Optimize the Model

Here I aim for >75% accuracy by:

1. Adding more neurons.

2. Adding another hidden layer.

3. Trying different activation functions.


In [94]:
# New optimized model
optimized_nn = tf.keras.models.Sequential()

# Adjusted layers
optimized_nn.add(tf.keras.layers.Dense(units=100, input_dim=number_input_features, activation="relu"))
optimized_nn.add(tf.keras.layers.Dense(units=50, activation="relu"))
optimized_nn.add(tf.keras.layers.Dense(units=20, activation="relu"))
optimized_nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Compile and train
optimized_nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])
optimized_nn.fit(X_train_scaled, y_train, epochs=150)

# Save optimized model
optimized_nn.save('AlphabetSoupCharity_Optimization.h5')

Epoch 1/150
804/804 [==============================] - 1s 980us/step - loss: 0.5700 - accuracy: 0.7217
Epoch 2/150
804/804 [==============================] - 1s 969us/step - loss: 0.5551 - accuracy: 0.7301
Epoch 3/150
804/804 [==============================] - 1s 969us/step - loss: 0.5527 - accuracy: 0.7310
Epoch 4/150
804/804 [==============================] - 1s 959us/step - loss: 0.5511 - accuracy: 0.7341
Epoch 5/150
804/804 [==============================] - 1s 904us/step - loss: 0.5491 - accuracy: 0.7336
Epoch 6/150
804/804 [==============================] - 1s 973us/step - loss: 0.5477 - accuracy: 0.7348
Epoch 7/150
804/804 [==============================] - 1s 983us/step - loss: 0.5479 - accuracy: 0.7345
Epoch 8/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5468 - accuracy: 0.7363
Epoch 9/150
804/804 [==============================] - 1s 971us/step - loss: 0.5464 - accuracy: 0.7357
Epoch 10/150
804/804 [==============================] - 1s 1ms/step - loss:

804/804 [==============================] - 1s 942us/step - loss: 0.5357 - accuracy: 0.7409
Epoch 81/150
804/804 [==============================] - 1s 951us/step - loss: 0.5354 - accuracy: 0.7409
Epoch 82/150
804/804 [==============================] - 1s 975us/step - loss: 0.5357 - accuracy: 0.7412
Epoch 83/150
804/804 [==============================] - 1s 955us/step - loss: 0.5353 - accuracy: 0.7413
Epoch 84/150
804/804 [==============================] - 1s 931us/step - loss: 0.5352 - accuracy: 0.7409
Epoch 85/150
804/804 [==============================] - 1s 955us/step - loss: 0.5354 - accuracy: 0.7413
Epoch 86/150
804/804 [==============================] - 1s 959us/step - loss: 0.5351 - accuracy: 0.7412
Epoch 87/150
804/804 [==============================] - 1s 966us/step - loss: 0.5351 - accuracy: 0.7416
Epoch 88/150
804/804 [==============================] - 1s 954us/step - loss: 0.5358 - accuracy: 0.7411
Epoch 89/150
804/804 [==============================] - 1s 959us/step - loss:

Step 4: Report

-Overview of the analysis.

-Preprocessing steps.

-Model architecture and accuracy.

-Optimization attempts.

In [99]:
# Define report content
report_content = f"""
# Deep Learning Challenge Report: Predicting Alphabet Soup Success

## **1. Overview of the Analysis**
Alphabet Soup is a nonprofit foundation that provides funding to organizations. The goal of this project is to build a deep learning model that predicts whether an applicant will successfully use the funding provided by Alphabet Soup. 

I use **machine learning and neural networks** to classify funding applicants based on historical data. The dataset contains over **34,000** organizations and includes categorical and numerical features related to the applicants.

---

## **2. Data Preprocessing**
To prepare the dataset for training, the following steps were performed:

- **Loaded the dataset** and explored its structure.
- **Dropped unnecessary columns:** The `EIN` and `NAME` columns, as they do not contribute to the prediction.
- **Categorical Encoding:**
  - Replaced rare categories (less than a certain threshold) in `APPLICATION_TYPE` and `CLASSIFICATION` with "Other."
  - Applied **one-hot encoding** to categorical variables.
- **Feature Selection:**
  - Extracted features (`X`) and target variable (`y`).
  - Split the dataset into **training (75%)** and **testing (25%)** sets.
- **Feature Scaling:**
  - Used **StandardScaler** to normalize numerical values.

---

## **3. Model Architecture and Accuracy**
We implemented a **deep neural network (DNN)** using TensorFlow and Keras. The model structure is as follows:

- **Input Layer:** `{len(X[0])}` features
- **Hidden Layer 1:** 80 neurons, **ReLU** activation
- **Hidden Layer 2:** 30 neurons, **ReLU** activation
- **Output Layer:** 1 neuron, **Sigmoid** activation

### **Model Training Results**
After training the model for **100 epochs**, we obtained the following results:

- **Loss:** {round(model_loss, 4)}
- **Accuracy:** {round(model_accuracy * 100, 2)}%

---

## **4. Optimization Attempts**
To improve model performance beyond **75% accuracy**, we made the following modifications:

1. **Increased the number of neurons per layer** (from 80 → 100, 30 → 50).
2. **Added an additional hidden layer** to improve feature extraction.
3. **Tried different activation functions** (ReLU, Sigmoid, Tanh).
4. **Extended training epochs** from 100 to 150.

The best-performing model achieved an accuracy of **{round(model_accuracy * 100, 2)}%**, which is {">= 75%" if model_accuracy >= 0.75 else "< 75%"}.

---

## **5. Conclusion**
- The final model provides a reasonable prediction accuracy of **{round(model_accuracy * 100, 2)}%**.
- Further improvements can be made by:
  - Experimenting with **hyperparameter tuning**.
  - Using **dropout layers** to prevent overfitting.
  - Exploring different **machine learning models (e.g., Random Forest, XGBoost)** as alternatives to deep learning.

The deep learning approach demonstrates **strong potential** for predicting successful funding applicants for Alphabet Soup. 🚀
"""

# Save the report as a markdown file
report_filename = "AlphabetSoup_DeepLearning_Report.md"

with open(report_filename, "w") as file:
    file.write(report_content)

print(f"Report successfully saved as {report_filename}")


Report successfully saved as AlphabetSoup_DeepLearning_Report.md


In [100]:
# Convert Markdown to HTML
with open("AlphabetSoup_DeepLearning_Report.md", "r") as file:
    md_content = file.read()

html_content = markdown.markdown(md_content)

# Save as an HTML file
with open("AlphabetSoup_DeepLearning_Report.html", "w") as file:
    file.write(html_content)

print("Report successfully converted to HTML.")


Report successfully converted to HTML.


Final Output
Markdown Report: AlphabetSoup_DeepLearning_Report.md
HTML Report: AlphabetSoup_DeepLearning_Report.html
PDF Report: AlphabetSoup_DeepLearning_Report.pdf

Deep Learning Challenge Report: Predicting Alphabet Soup Success

1. Overview of the Analysis

Alphabet Soup is a nonprofit foundation that provides funding to organizations. The goal of this project is to build a deep learning model that predicts whether an applicant will successfully use the funding provided by Alphabet Soup.

I use machine learning and neural networks to classify funding applicants based on historical data. The dataset contains over 34,000 organizations and includes categorical and numerical features related to the applicants.

2. Data Preprocessing

To prepare the dataset for training, the following steps were performed:

Loaded the dataset and explored its structure.
Dropped unnecessary columns: The EIN and NAME columns, as they do not contribute to the prediction.
Categorical Encoding:
Replaced rare categories (less than a certain threshold) in APPLICATION_TYPE and CLASSIFICATION with "Other."
Applied one-hot encoding to categorical variables.
Feature Selection:
Extracted features (X) and target variable (y).
Split the dataset into training (75%) and testing (25%) sets.
Feature Scaling:
Used StandardScaler to normalize numerical values.
3. Model Architecture and Accuracy

We implemented a deep neural network (DNN) using TensorFlow and Keras. The model structure is as follows:

Input Layer: 43 features
Hidden Layer 1: 80 neurons, ReLU activation
Hidden Layer 2: 30 neurons, ReLU activation
Output Layer: 1 neuron, Sigmoid activation
Model Training Results

After training the model for 100 epochs, we obtained the following results:

Loss: 0.5569
Accuracy: 72.47%


4. Optimization Attempts

To improve model performance beyond 75% accuracy, we made the following modifications:

Increased the number of neurons per layer (from 80 → 100, 30 → 50).
Added an additional hidden layer to improve feature extraction.
Tried different activation functions (ReLU, Sigmoid, Tanh).
Extended training epochs from 100 to 150.
The best-performing model achieved an accuracy of 72.47%, which is < 75%.

5. Conclusion

The final model provides a reasonable prediction accuracy of 72.47%.
Further improvements can be made by:
Experimenting with hyperparameter tuning.
Using dropout layers to prevent overfitting.
Exploring different machine learning models (e.g., Random Forest, XGBoost) as alternatives to deep learning.
The deep learning approach demonstrates strong potential for predicting successful funding applicants for Alphabet Soup.